### ISSUE TITLE
[Proposal] Plan for VLM export support starting with Qwen3-VL

### ISSUE THREAD

**Issue by sinitame** [sonos/torch-to-nnef#68]
## Description
I would like to propose a small sequence of changes to introduce VLM export support, starting with the Qwen3-VL family.

The main idea is to keep the current export flow generic, while moving architecture-specific logic behind dedicated handlers. This should make it easier to support multimodal models without adding too much model-specific branching in the exporter.
For VLMs, the decoder needs more than input_ids and KV cache. It may also depend on visual embeddings and additional multimodal state. Keeping those details in handlers should help preserve a simpler exporter core.

### Proposed plan
1. Introduce a model-handler layer for architecture specific loading, wrapper selection, and input preparation.
2. Add initial Qwen3-VL decoder export support.
3. Export the vision encoder separately from the decoder.

### Initial goal
The initial target would be support for Qwen3-VL, with the longer-term goal of making the export path easier to extend to other multimodal families.

**Comment by JulienBalianSonos** [comment]
Hi @sinitame, thanks a lot for opening this and for the PR! Really happy to see contribution on this front, especially since the staged plan you propose lines up directly with what we had 'poorly' sketched in #17.
 
The split (handler layer -> Qwen3-VL decoder -> vision encoder separately) makes sense to me, and keeping arch-specific logic behind dedicated handlers is the right shape for where we want the LLM exporter to go.

I will review #67 during the week. Looking forward to the Qwen3-VL follow-up too. If you hit any blockers along the way (tract export quirks, KV cache edge cases, anything in the existing wrapper layer), please ping me here or on the PRs -- happy to help unblock so we can iterate quickly toward the VLM stages.

**Comment by JulienBalianSonos** [comment]
How do you see the exported IO surface evolving past stage 1?

 Today the wrapper signature is hardcoded to (input_ids, kv_*) via update_forward_signature (reading from `HFConfigHelper.build_kv_cache_infos`), and the return is fixed at [logits, *kvs]. As soon as users want to expose more in the NNEF graph, the wrapper becomes the bottleneck, not the handler.

Concrete cases:
- External attention_mask (padded batches, prefix-only) instead of internal triu.
- position_ids / cache_position as inputs (YaRN, sliding windows, resumed generation).
- inputs_embeds as input. Big one for VLMs: pre-compute image embeddings outside the graph; also soft prompts.
- VLM-native inputs: pixel_values, image_grid_thw, cross_attention_states, audio features.
- Per-token logits or hidden states as outputs (perplexity, retrieval, classifier heads).

A handler-driven design covers this if the handler owns both the call and the IO:
```python
  class ArchitectureHandler:
      def build_input_spec(...) -> InputSpec                       # already in this PR
      def call_model(self, model, inputs) -> tuple[Tensor, ...]    # matches output_names
```

`update_forward_signature` builds the wrapper signature from input_spec.input_names instead of hardcoding (input_ids, kv_*), and the return is packed from call_model matching output_names. A Qwen3-VL handler can then declare (input_ids, pixel_values, image_grid_thw, kv_*) -> (logits, kvs, ...), no wrapper change needed.

This pairs with the wrapper-unification point above: collapsing to one wrapper only pays off if the handler can fully describe its IO, otherwise we just trade one rigid wrapper for another.

Worth settling before stage 2 since Qwen3-VL is the first arch where this bites. Curious what you had in mind here.

**Comment by sinitame** [comment]
My initial attempt was indeed a bit of a bypass. On the first Qwen3-VL pass, I extended the wrapper with:

  - `handler: T.Optional[ArchitectureHandler] = None`, so the wrapper could delegate architecture-specific input preparation
  - `num_kv_tensors: T.Optional[int] = None`, so it could still distinguish KV-cache tensors from any extra VLM state tensors

  Inside `forward()`, I was planning to add VLM-specific arguments through the existing positional wrapper surface with:
```python
  inputs = (input_ids, *args)
  model_inputs = self.handler.prepare_inputs_for_model(
      inputs=inputs,
      wrapper=self,
  )
```

That was enough to make the first VLM attempt work, but it also confirmed your point: the wrapper was still the real owner of the IO shape, and the handler was adapting to it rather than describing it.

Following the review, I aligned the design more with what you suggest. The current direction is to make the handler responsible for the IO description and the full forward boundary:
```python
  class ArchitectureHandler:
      def build_io_spec(...) -> IOSpec
      def build_forward_inputs(..., exported_inputs_by_name) -> dict[str, Any]
      def call_model(..., model, prepared_inputs) -> Any
      def build_forward_outputs(..., model_outputs, prepared_inputs) -> tuple[Tensor, ...]
```

 As you proposed, I think a next step in stage 2 will be to change the directly with something like this (instead of doing the input additions in the forward pass directly):
```python
  def update_forward_signature(self, input_spec: InputSpec):
      sign = inspect.signature(self.forward)
      kind = inspect.Parameter.POSITIONAL_OR_KEYWORD
      new_params = OrderedDict(
          (
              name,
              inspect.Parameter(name, kind, annotation=torch.Tensor),
          )
          for name in input_spec.input_names
      )
      ...
```

That would let a Qwen3-VL handler declare its full surface directly, instead of relying on the wrapper to interpret positional additions after the KV tensors.
